### PR TITLE
Fix validation issue with (dynamically) optional field

### PR DESCRIPTION
**Pull request by sergei-maertens**
Taiga DH 671

~~If the test suite passes first try I'm eating my shoe.~~

Needs backport to stable/2.6.x

### CODE REVIEW

File: src/openforms/submissions/form_logic.py [hunk 1]
```diff
@@ -214,15 +208,19 @@ def check_submission_logic(
         submission_variables_state.set_values(unsaved_data)
     data_container = DataContainer(state=submission_variables_state)
 
-    mutation_operations = []
+    mutation_operations: list[ActionOperation] = []
     for operation in iter_evaluate_rules(rules, data_container, submission):
```
Comment (Viicos) [inline src/openforms/submissions/form_logic.py:212]: Probably this can be built using a list comprehension now

**Review by Viicos — APPROVED** [review]
So if I understand correctly this was a bug before backend validation, but only appears now with this validation?

**Comment by sergei-maertens** [comment]
> So if I understand correctly this was a bug before backend validation, but only appears now with this validation?

No, because the logic evaluation for the entire submission runs after all form steps have been submitted. It was deliberate to skip the property actions, since they affect the UI/client side validation, but that was not presented anymore because the user is on the "confirmation" screen.

File: src/openforms/submissions/form_logic.py (hunk 1 above)
Comment (sergei-maertens) [reply src/openforms/submissions/form_logic.py:212]: I think a list comprehension makes the code look pure, but it isn't, data mutations are applied as the logic rules are evaluated by looping through them.

**Comment by sergei-maertens** [comment]
Backported to stable/2.6.x in b0dc21dca80b6bbefaefdfd956a793e85ee721dd

**Pull request merged**